js/adapt-stateful-session.js

```diff
@@ -1,227 +1,259 @@
+      // when the user switches language, we need to:
+      // - reattach the event listeners as the language change triggers a reload of
+      //   the json, which will create brand new collections
+      // - get and save a fresh copy of the session state. as the json has been reloaded,
+      //   the blocks completion data will be reset (the user is warned that this will
+      //   happen by the language picker extension)
+      // - check to see if the config requires that the lesson_status be reset to
+      //   'incomplete'
+      const config = Adapt.spoor.config;
       this.removeEventListeners();
       this.setupEventListeners();
-    },
-
-    onBlockComplete: function(block) {
       this.saveSessionState();
-    },
+      if (config._reporting && config._reporting._resetStatusOnLanguageChange === true) {
+        Adapt.offlineStorage.set('status', 'incomplete');
+      }
+    }
 
-    onTrackingComplete: function(completionData) {
-      this.saveSessionState();
+    onVisibilityChange() {
+      if (document.visibilityState === 'hidden') this.scorm.commit();
+    }
 
-      var completionStatus = completionData.status.asLowerCase;
+    onQuestionRecordInteraction(questionView) {
+      if (!this._shouldRecordInteractions) return;
+      const responseType = questionView.getResponseType();
+      // If responseType doesn't contain any data, assume that the question
+      // component hasn't been set up for cmi.interaction tracking
+      if (_.isEmpty(responseType)) return;
+      const id = questionView.model.get('_id');
+      const response = questionView.getResponse();
+      const result = questionView.isCorrect();
+      const latency = questionView.getLatency();
+      Adapt.offlineStorage.set('interaction', id, response, result, latency, responseType);
+    }
 
+    onAssessmentComplete(stateModel) {
+      const config = Adapt.spoor.config;
+      Adapt.course.set('_isAssessmentPassed', stateModel.isPass);
+      this.saveSessionState();
+      const shouldSubmitScore = (config && config._tracking && config._tracking._shouldSubmitScore);
+      if (!shouldSubmitScore) return;
+      const scoreArgs = stateModel.isPercentageBased ?
+        [ stateModel.scoreAsPercent, 0, 100 ] :
+        [ stateModel.score, 0, stateModel.maxScore ];
+      Adapt.offlineStorage.set('score', ...scoreArgs);
+    }
+
+    onTrackingComplete(completionData) {
+      const config = Adapt.spoor.config;
+      this.saveSessionState();
+      let completionStatus = completionData.status.asLowerCase;
       // The config allows the user to override the completion state.
       switch (completionData.status) {
         case COMPLETION_STATE.COMPLETED:
         case COMPLETION_STATE.PASSED: {
-          if (!this._config._reporting._onTrackingCriteriaMet) {
-            Adapt.log.warn("No value defined for '_onTrackingCriteriaMet', so defaulting to '" + completionStatus + "'");
+          if (!config._reporting._onTrackingCriteriaMet) {
```

Comment:
Does `config` have to be checked for a truthy around here?

Reply:
done in https://github.com/adaptlearning/adapt-contrib-spoor/pull/197/commits/6fe523b804c56521147c4358729dd9fcc2c6fb11